### PR TITLE
PiFi-40 DeviceTree overlay added

### PR DESCRIPTION
**Pull request by pifi-bz**
Hi - have another audio device to add to the RPi kernel.  I've separated the requests into three as per the recent request to Allo, and here's the first one - I thought I'd just submit this on its own to ensure that I've the basic process correct.

### CODE REVIEW

File: arch/arm/boot/dts/overlays/README [hunk 1]
```diff
@@ -2000,6 +2000,10 @@ Params: spi-present-mask        8-bit integer, bitmap indicating MCP23S17 SPI0
                                 0-3, which can be configured with JP1 and JP2.
 
 
+Name:	pifi-40
+Info:	Configures the PiFi 40W stereo amplifier
+Load:	dtoverlay=pifi-40
+
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/README:2008]: The formatting rules for the README are very strict:
* No hard TABs - indent with spaces, to column 8 or 32 (or 9/33 if numbering from 1) 
* No trailing whitespace
* All entries must have Name, Info, Load and Params.
* 2 blank lines between entries

File: arch/arm/boot/dts/overlays/pifi-40-overlay.dts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+				sound-name-prefix = "Right";
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@2 {
+		target = <&sound>;
+		pifi_40: __overlay__ {
+			compatible = "pifi,pifi-40";
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/pifi-40-overlay.dts:43]: You appear to have a dedicated board driver - I'm not going to merge the overlay without it.

File: arch/arm/boot/dts/overlays/pifi-40-overlay.dts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+	fragment@1 {
+		target = <&i2c1>;
+		__overlay__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			tas5711l: audio-codec@1a {
+				compatible = "ti,tas5711";
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/pifi-40-overlay.dts:23]: This driver isn't enabled in current builds.

File: arch/arm/boot/dts/overlays/pifi-40-overlay.dts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+		};
+	};
+
+	fragment@2 {
+		target = <&sound>;
+		pifi_40: __overlay__ {
+			compatible = "pifi,pifi-40";
+			audio-codec = <&tas5711l &tas5711r>;
+			i2s-controller = <&i2s>;
+			pdn-gpios = <&gpio 23 1>;
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/pifi-40-overlay.dts:46]: Is this really active low? Be aware that GPIO23 (like GPIOs 9-27) pulls down by default and will remain low (and therefore "active") until the driver starts and sets it to the preferred value (presumably high/inactive).

**Review by pelwell — COMMENTED** [review]
Ah - I see you've added a config setting to the defconfigs for the driver, but it's still lacking the driver itself and the config setting for the codec.

File: arch/arm/boot/dts/overlays/pifi-40-overlay.dts (hunk 4 above)
Comment (pifi-bz) [reply arch/arm/boot/dts/overlays/pifi-40-overlay.dts:46]: Yep - it's an active-low "power down" input.

File: sound/soc/bcm/Makefile [hunk 5]
```diff
@@ -69,3 +70,4 @@ obj-$(CONFIG_SND_PISOUND) += snd-soc-pisound.o
 obj-$(CONFIG_SND_BCM2708_SOC_FE_PI_AUDIO) += snd-soc-fe-pi-audio.o
 obj-$(CONFIG_SND_RPI_SIMPLE_SOUNDCARD) += snd-soc-rpi-simple-soundcard.o
 obj-$(CONFIG_SND_RPI_WM8804_SOUNDCARD) += snd-soc-rpi-wm8804-soundcard.o
+obj-$(CONFIG_SND_BCM2708_SOC_PIFI_40) += snd-soc-pifi-40.o
```
Comment (pelwell) [inline sound/soc/bcm/Makefile]: The GitHub GUI is pointing out the lack of a newline at the end of this file.

File: sound/soc/bcm/pifi-40.c [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+	"Speaker Switch"
+};
+
+static int snd_pifi_40_init(struct snd_soc_pcm_runtime *rtd)
+{
+	struct snd_soc_card *card = rtd->card;
+	struct snd_soc_component *dac[2];
+	struct snd_kcontrol *kctl;
+	int i, j;
+	dac[0] = rtd->codec_dais[0]->component;
```
Comment (pelwell) [inline sound/soc/bcm/pifi-40.c:85]: Kernel coding style is to have a blank line between variable declarations and code.

File: sound/soc/bcm/pifi-40.c [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+	gpiod_set_value_cansleep(pdn_gpio, 0);
+	usleep_range(20000, 30000);
+
+	// Oscillator trim
+	snd_soc_component_write(dac[0], 0x1b, 0);
+	snd_soc_component_write(dac[1], 0x1b, 0);
+	usleep_range(60000, 80000);
+
+	// Common setup
+	for (i=0; i<2; i++) {
```
Comment (pelwell) [inline sound/soc/bcm/pifi-40.c]: And spaces around operators - `for (i = 0; i < 2; i++) {`

File: arch/arm/configs/bcm2709_defconfig [hunk 8]
```diff
@@ -1019,6 +1020,7 @@ CONFIG_SND_SOC_ICS43432=m
 CONFIG_SND_SOC_MA120X0P=m
 CONFIG_SND_SOC_MAX98357A=m
 CONFIG_SND_SOC_SPDIF=m
+CONFIG_SND_SOC_TAS571X=m
```
Comment (pelwell) [inline arch/arm/configs/bcm2709_defconfig]: The `select SND_SOC_TAS571X` in your `SND_BCM2708_SOC_PIFI_40` declaration means it is not necessary to explicitly include this setting - apologies if the chronology of your updates means I could have spotted this sooner.

File: arch/arm/boot/dts/overlays/README (hunk 1 above)
Comment (pelwell) [reply arch/arm/boot/dts/overlays/README:2008]: It's still short a blank line after this addition.

File: arch/arm/configs/bcm2709_defconfig [hunk 9]
```diff
@@ -984,6 +984,7 @@ CONFIG_SND_BCM2708_SOC_HIFIBERRY_DACPLUSADCPRO=m
 CONFIG_SND_BCM2708_SOC_HIFIBERRY_DACPLUSDSP=m
 CONFIG_SND_BCM2708_SOC_HIFIBERRY_DIGI=m
 CONFIG_SND_BCM2708_SOC_HIFIBERRY_AMP=m
+CONFIG_SND_BCM2708_SOC_PIFI_40=m
```
Comment (pelwell) [inline arch/arm/configs/bcm2709_defconfig:987]: bcmrpi_defconfig also needs this setting.

**Review by pelwell — COMMENTED** [review]
There are a few minor inline comments to address. After that, squash your changes into three commits:
1. the driver and its Makefile and Kconfig changes,
2. the overlay, Makefile and README, and
3. the defconfig changes.

These commits should be run through scripts/checkpatch.pl to spot style problem - even downstream we expect people to follow the kernel coding conventions, but we don't demand that compatible strings are documented and the MAINTAINER file is updated. `Signed-off-by` lines are required though.

File: arch/arm/configs/bcm2709_defconfig (hunk 8 above)
Comment (pifi-bz) [reply arch/arm/configs/bcm2709_defconfig]: Any preference as to whether I should leave them in or take them out?  Just spotted that the TAS571X module's not in bcmrpi_defconfig but it is in all the rest; I guess it should be consistent..

File: arch/arm/configs/bcm2709_defconfig (hunk 9 above)
Comment (pifi-bz) [reply arch/arm/configs/bcm2709_defconfig:987]: Added.

File: sound/soc/bcm/pifi-40.c (hunk 7 above)
Comment (pifi-bz) [reply sound/soc/bcm/pifi-40.c]: Ran it through clang-format and then scripts/checkpatch.pl.

Just want to add a note of thanks for your taking the time to point all of these things out.  It's much appreciated, and I'm taking this all on board for the next time.

File: arch/arm/configs/bcm2709_defconfig (hunk 8 above)
Comment (pelwell) [reply arch/arm/configs/bcm2709_defconfig]: Take them out - periodically we use "make savedefconfig", and that will remove any setting which is implied by another.

**Comment by pelwell** [comment]
That all looks good. Are you up for doing the squashes? Unless you are good at git commit juggling you might find it easiest to take a snapshot of your current state ("git branch snapshot"), then "git reset HEAD~12" (assuming you've still got 12 commits) and add the files again in the groupings mentioned above to form three new commits. Don't forget to use "git commit -s" to get the "Signed-off-by:" attribution.

File: arch/arm/configs/bcm2709_defconfig (hunk 8 above)
Comment (pifi-bz) [reply arch/arm/configs/bcm2709_defconfig]: Done - thanks!

**Pull request merged**